include/rosparam_shortcuts/parameter.h

```diff
@@ -0,0 +1,162 @@
+#pragma once
+
+#include <mutex>
+#include <shared_mutex>
+
+#include <rclcpp>
+
+namespace rosparam_shortcuts
+
+    /*
+     * Thread-safe container for generic parameter values, providing callback functions.
+     * This class implements a tree structure of parameters which are declared recursively.
+     * All leafs are of type Parameter<rclcpp::Parameter> which are used for interacting with
+     * the rclcpp node parameter interface using the NodeConfig class further down.
+     */
+    template <typename ParameterT>
+    class Parameter
+{
+public:
+	Parameter(const std::string& name);
+	Parameter(const std::string& name, const ParameterT& default_value);
+
+	// Fields
+	const std::string& name() const noexcept;
+	std::string description() const noexcept;
+	std::string additional_constraints() const noexcept;
+	bool isNodeParameter() const noexcept;  // true if this parameter contains an rclcpp::Parameter
+
+	// Properties
+	void describe(const std::string& description, const std::string& additional_constraints = "");
+	// Allow setting additional properties. i.e. read_only, dynamic_typing?
+	// void setProperties(...);
+
+	// Validity Status
+	const bool isValid() const noexcept;  // errors empty?
+	const bool hasErrors() const noexcept;
+	const bool hasWarnings() const noexcept;
+	// TODO: Insert additional thread-safe access to errors/warnings
+
+	// Read/write-locked data access using std::shared_mutex
+	// See: https://www.youtube.com/watch?v=ozOgzlxIsdg
+	ReadLockedPtr<ParameterT> operator->const;
+	WriteLockedPtr<ParameterT> operator->;
+	// optional: (useful for synchronized vector/map operations)
+	void applyRW(std::function<void(ParameterT& value)>);
+	void applyRO(std::function<void(const ParameterT& value)>);
+
+	// Copy-set/Copy-get value using read/write lock internally
+	ParameterT& operator=(const ParameterT& value);
+	ParameterT operator() const;  // or:  'operator ParameterT() const;'
+
+	// Validity Callbacks (could also use rcl_interfaces::msg::SetParametersResult like in node_parameters.h).
```

Comment:
There could also be something like `bool testValueChange(const ParameterT& new_value)' which triggers the same (chained) validity callbacks only without actually changing the value. Incoming parameter changes from other nodes can be rejected if they are not valid.

Reply:
That seems useful.  It could also mean that subsystems inside the same node could internally "test a config change" although I'm not sure that's a pattern we want to encourage.  There are better ways for a node to communicate internally.

Reply:
I'd argue that a node should probably not change its own parameters internally, except for setting defaults maybe. I mean, the use case of dynamic parameters is that _other_ nodes (e.g. user interfaces) can change the modality of the nodes subsystem.